### PR TITLE
Add a collapsible group for installation preliminaries #67

### PR DESCRIPTION
**Pull request by nedbat**
This accomplishes what I wanted, though I'm not sure about the code.  I used a global because I didn't know what was allowed for storing state in the config and arguments.  Let me know if there's a better way to do it.

### CODE REVIEW

**Review by gaborbernat — APPROVED** [review]

File: src/tox_gh/plugin.py [hunk 1]
```diff
@@ -80,14 +80,37 @@ def tox_add_core_config(core_conf: ConfigSet, state: State) -> None:
...
+def tox_on_install(tox_env: ToxEnv, arguments: Any, section: str, of_type: str) -> None:  # noqa: ANN401, ARG001
+    """
+    Run before installing to prepare an environment.
+
+    :param tox_env: the tox environment
+    :param arguments: installation arguments
+    :param section: section of the installation
+    :param of_type: type of the installation
+    """
+    global installing  # noqa: PLW0603
```
Comment (gaborbernat) [inline src/tox_gh/plugin.py:96]: We should store state in the environment via tox_env https://tox.wiki/en/4.6.4/plugins_api.html#tox.config.sets.ConfigSet.add_constant

**Review by gaborbernat — CHANGES_REQUESTED** [review]
Let's remove the global.

File: src/tox_gh/plugin.py (hunk 1 above)
Comment (nedbat) [reply src/tox_gh/plugin.py:96]: I saw `add_constant`, but this isn't a constant, it's value changes.  Maybe I don't understand what constant means here?  Can you point me to an example that stores state in the way you like? I tried searching GitHub.

Comment (gaborbernat) [reply src/tox_gh/plugin.py:96]: It is a constant no? You set it once? A constant can have two states unset and set. (it's a different conversation what its value is).

Comment (nedbat) [reply src/tox_gh/plugin.py:96]: It flips between true (line 98) and false (line 112). It's not configuration, so it seems odd to have to pick a key and a description for it.  The description for `add_constant` says, "Add a constant value."

Comment (gaborbernat) [reply src/tox_gh/plugin.py:96]: Sounds to me like `add_constant("install_triggered", True)`

**Comment by nedbat** [comment]
As the name seems to imply, I cannot change the value:
```
  File "/Users/runner/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/tox_gh/plugin.py", line 95, in tox_on_install
    tox_env.core.add_constant(keys="in_installation_phase", desc="flip-flop for installation", value=True)
  File "/Users/runner/hostedtoolcache/Python/3.8.17/x64/lib/python3.8/site-packages/tox/config/sets.py", line 83, in add_constant
    raise RuntimeError(msg)
RuntimeError: config set has been marked final and cannot be extended
```

I need a boolean value that I can start as False, then switch between True and False, two switches for each environment run, as tox moves from installation to execution and back.

**Comment by gaborbernat** [comment]
> I need a boolean value that I can start as False, then switch between True and False, two switches for each environment run, as tox moves from installation to execution and back.

Ah in that case it's not a constant just a normal configuration value 👍

**Comment by nedbat** [comment]
> Ah in that case it's not a constant just a normal configuration value 👍

This still seems really strange to me: it's not a configuration value at all.  It's just state of the plugin.

When I try to use `add_config` to change the value:
```
E           File "/System/Volumes/Data/root/src/tox-gh/.tox/py311/lib/python3.11/site-packages/tox_gh/plugin.py", line 95, in tox_on_install
E             tox_env.core.add_config(keys="in_installation_phase", of_type=bool, default=True, desc="flip-flop")
E           File "/System/Volumes/Data/root/src/tox-gh/.tox/py311/lib/python3.11/site-packages/tox/config/sets.py", line 66, in add_config
E             raise RuntimeError(msg)
E         RuntimeError: config set has been marked final and cannot be extended
```

I tried setting the value:
```
E           File "/System/Volumes/Data/root/src/tox-gh/.tox/py311/lib/python3.11/site-packages/tox_gh/plugin.py", line 95, in tox_on_install
E             tox_env.core["in_installation_phase"] = True
E             ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
E         TypeError: 'CoreConfigSet' object does not support item assignment
```

This is frustrating.

**Comment by gaborbernat** [comment]
I mean can be a global too, will just not work in parallel mode 🤔 and is kinda ugly. Likely we should have better mechanism for such transient state 🤔

**Comment by nedbat** [comment]
If the plugin were a class, would it work as an instance attribute?

**Comment by gaborbernat** [comment]
> If the plugin were a class, would it work as an instance attribute?

Plugins are not classes though. If you revert it to the global variable, I'll accept it.

**Comment by nedbat** [comment]
I've changed it to a thread local, to hopefully accommodate tox run-parallel.

File: src/tox_gh/plugin.py [hunk 2]
```diff
@@ -80,6 +81,26 @@ def tox_add_core_config(core_conf: ConfigSet, state: State) -> None:
         state.conf.core.loaders.insert(0, MemoryLoader(env_list=env_list))
 
 
+state = threading.local()
```
Comment (gaborbernat) [inline src/tox_gh/plugin.py:84]: This should be _STATE because is global and private really.

**Review by gaborbernat — CHANGES_REQUESTED** [review]

**Comment by nedbat** [comment]
OK, I think I've done everything I can.  The output is still a little odd ([coverage.py action run](https://github.com/nedbat/coveragepy/actions/runs/5636010187/job/15267624871)):
```
▶ Run python -m tox -- -rfsEX
ROOT: running tox-gh
ROOT: tox-gh set py311, py310
▶ tox:install
▶ tox:py311
py311: OK ✔ in 2 minutes 17.12 seconds
▶ tox:install
▶ tox:py310
.pkg: _exit> python /opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg-cpython310: _exit> python /opt/hostedtoolcache/Python/3.11.4/x64/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  py311: OK (137.12=setup[18.99]+cmd[0.39,0.96,5.18,60.06,0.17,51.37] seconds)
  py310: OK (124.57=setup[18.70]+cmd[0.33,0.94,5.31,51.62,0.18,47.50] seconds)
  congratulations :) (263.72 seconds)
```
Not sure why there's a "py311: OK in 2 minutes" line when there isn't one for py310 (even in the collapsed section).

I'll make another PR to move the "ROOT" lines from a warning to a lower logging level.

**Comment by gaborbernat** [comment]
That's expected see https://github.com/tox-dev/tox/issues/3068

**Comment by gaborbernat** [comment]
The CI is failing can you please fix that 😊

**Comment by nedbat** [comment]
The CI is failing because the tests don't run more than one installation step before running the commands.  I'm not sure how to get them to do that.  What can I add to the ini in the tests to do that?

**Comment by gaborbernat** [comment]
deps + package would be 2 🤔

**Comment by nedbat** [comment]
Done, though it was complicated.

**Pull request merged**